src/RestQuery.js

```diff
@@ -752,22 +752,20 @@ RestQuery.prototype.handleInclude = function() {
     return;
   }
 
-  var pathResponse = includePath(
-    this.config,
-    this.auth,
-    this.response,
-    this.include[0],
-    this.restOptions
-  );
-  if (pathResponse.then) {
-    return pathResponse.then(() => {
-      this.include = this.include.slice(1);
-      return this.handleInclude();
-    });
-  } else if (this.include.length > 0) {
-    this.include = this.include.slice(1);
-    return this.handleInclude();
-  }
+  // Construct a graph of promises, ensuring that we don't try to load
+  // any path until it's prefix has finished loading.
+  const promises = { '': Promise.resolve() };
```

Comment:
I don't understand this idiom.  trying to figure out how i can figure it out :).

Reply:
This is a little bit tricky, I could probably comment this more clearly!

What this is intending is that entries in `this.includes` that have no nesting will start processing right away, by chaining a `then` handler on to this resolved promise.

For an include with no nesting, its entry in `this.include` will be an array with one element: `['user']`, for instance. From that path we extract a `prefix`, which in this case will be `[].join('.')` which is the empty string `''`.  We would find the above base case with `promises['']`.

It might be even clearer to write this as an inline default:
```js
const promises = {};
this.include.forEach(path => {
  const prefix = path.slice(0, -1).join('.');
  const loadAfter = promises[prefix] || Promise.resolve();
  ...
}
```

That seems like it is more idiomatic JS, I think I'll probably follow up and switch it over.